src/abstractarray.jl

```diff
@@ -291,3 +291,22 @@ function Base.view(S::SArray, I::Union{Colon, Integer, SOneTo, StaticArray{<:Tup
     V = getindex(S, I...)
     _maybewrapscalar(S, V)
 end
+
+# zeros, ones and fill may return SArrays if all the axes are statically sized
+for (cf, elf) in ((:zeros, :zero), (:ones, :one))
+    _cf = Symbol(:_, cf)
+    @eval begin
+        $_cf(::Type{T}, s::Size) where {T} = similar_type(SArray, T, s)(ntuple(_->$elf(T), prod(s)))
+        # fall back to Base if the size is not statically known
+        $_cf(T, sz) = $cf(T, sz)
+        function $cf(::Type{T}, ax::NTuple{N,HeterogeneousShape}) where {T,N}
+            $_cf(T, homogenize_shape(ax))
+        end
+    end
+end
+
+_fill(v, s::Size) = similar_type(SArray, typeof(v), s)(ntuple(_->v, prod(s)))
+_fill(v, sz) = fill(v, sz)
+function fill(v, ax::NTuple{N,HeterogeneousShape}) where {N}
```

Comment:
I think it would be fine to just define it for `NTuple` of `SOneTo`. Then there would be no (weak) type piracy. And similarly for other functions.

Reply:
The combination lets one address mixed argument types, although I'm unsure if it's a practical need. The weak type-piracy is unfortunate, but ultimately harmless as things stand currently. However, it's probably fine to only define it for `SOneTo`

Reply:
We don't support hybrid arrays in StaticArrays.jl so it would be something more appropriate for [HybridArrays.jl](https://github.com/JuliaArrays/HybridArrays.jl).

Reply:
Updated to remove the type-piracy.